### ISSUE TITLE
NamedTuple elements have Any type when indexed (and other tuple-related things)

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#860]
**Describe the bug**
(I acknowledge this is a bit of a hotchpotch but these are all issues I encountered when attempting to use tuples and named tuples from the `typing` module.)  True to the title of the bug report, Pyright is not able to determine the types of `NamedTuple` elements when indexing or slicing the `NamedTuple` as opposed to accessing elements by name.  More generally, Pyright is not able to capture the types of individual elements (nor the length of the resulting tuple) in a heterogeneous tuple when slicing.  Values are represented using `Literal`s for tuple literals but for `NamedTuples` only the type is shown.  Finally, though the square bracket notation is syntactic sugar for `__getitem__`, the types derived from `__getitem__` are generic to the tuple - I assume there is no special handling of `__getitem__` on top of what's provided by the typeshed.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
* That the type of `named_tuple[0]` would be `str` or a `Literal['']` rather than `Any` 
* That the type of `named_tuple[:2]` would be `Tuple[str, int]` or `Tuple[Literal[''], Literal[0]]` rather than `Tuple[Any, ...]` (an arbitrary-length tuple)
* That the type of `vanilla_tuple[:2]` would be `Tuple[Literal[''], Literal[0]]` rather than `Tuple[Literal['', 0, True], ...]`
* That the type of `vanilla_tuple.__getitem__(0)` would be `Literal['']` rather than `Literal['', 0, True]`
* That the type of `vanilla_tuple.__getitem__(slice(0, 2))` would be `Tuple[Literal[''], Literal[0]]` rather than `Tuple[Literal['', 0, True], ...]`

**Screenshots or Code**

```py
# pyright: strict

from typing import NamedTuple


class SomeNamedTuple(NamedTuple):
    a: str
    b: int
    c: bool

named_tuple = SomeNamedTuple('', 0, True)
reveal_type(named_tuple.a)
reveal_type(named_tuple[0])
reveal_type(named_tuple[:2])

vanilla_tuple = ('', 0, True)
reveal_type(vanilla_tuple[0])
reveal_type(vanilla_tuple[:2])
reveal_type(vanilla_tuple.__getitem__(0))
reveal_type(vanilla_tuple.__getitem__(slice(0, 2)))
```

**VS Code extension or command-line**
Extension, 1.1.54

**Comment by erictraut** [comment]
I don't consider any of these observations "bugs". They're feature requests. These all require Pyright to encode specialized knowledge of the internal workings of the NamedTuple class. I've generally tried to avoid doing this practice, instead relying on behaviors encoded in the typeshed type stub files. However, I think there's a reasonable justification for implemented some or all of the suggestions here.

**Comment by layday** [comment]
That's fair - I think it was mostly that I found the bug report layout more convenient.

**Comment by erictraut** [comment]
I've thought about this more, and I don't think I will add any of this support. As I mentioned above, it would require the type checker to make a bunch of assumptions about the internal workings of the NamedTuple and Tuple classes. I've tried hard to avoid this practice. Operations such as indexing should be described in the stdlib type stubs. Pyright does already support the case where the type being indexed is a Tuple[_T] and the index is an expression consisting of a constant integer. I felt this was justified because it's a simple index into the type arguments for the tuple. Adding support for any of the other cases above would require internal knowledge of the Tuple or NamedTuple classes.

**Comment by erictraut** [comment]
I've updated pyright to include support for unpacking and indexing of instances of NamedTuple or classes derived from NamedTuple. The indexing support works only in the case where the index expression is a simple integer numeric literal. This support will be in the next version.
